File: typescript/packages/metadata/src/loader/meta-data-source.ts
// MetaDataSource — the raw-document unit consumed by the loader pipeline.
//
// A loader (FileMetaDataLoader, later UrlMetaDataLoader) discovers/acquires
// sources; the MetaDataLoader pipeline calls read() on each. read() is async
// so file/URL sources can do I/O; InMemorySource resolves immediately.

/** Format of a source's content. Selects the parser. */
export type MetaDataFormat = "json" | "yaml";

/** One unit of raw metadata input. */
export interface MetaDataSource {
  /** Human-readable identifier — used in parse-error messages (e.g. a filename). */
  readonly id: string;
  /** Content format hint — selects the parser. */
  readonly format: MetaDataFormat;
  /** Resolve the raw content. May perform I/O. */
  read(): Promise<string>;
}

/** A metadata source backed by an in-memory string. */
export class InMemorySource implements MetaDataSource {
  readonly id: string;
  readonly format: MetaDataFormat;
  private readonly _content: string;

  constructor(content: string, opts?: { id?: string; format?: MetaDataFormat }) {
    this._content = content;
    this.id = opts?.id ?? "<in-memory>";
    this.format = opts?.format ?? "json";
  }

  read(): Promise<string> {
    return Promise.resolve(this._content);
  }
}

File: server/csharp/MetaObjects/Loader/MetaDataSource.cs
// MetaDataSource — the raw-document unit consumed by the loader pipeline.
//
// Ported from typescript/packages/metadata/src/loader/meta-data-source.ts
//
// The C# loader is synchronous (corpus is tiny; sync I/O keeps the adapter
// simple). Read() returns string, not Task<string>.

namespace MetaObjects.Loader;

/// <summary>Content format of a source — selects the parser.</summary>
public enum MetaDataFormat { Json, Yaml }

/// <summary>One unit of raw metadata input consumed by the loader pipeline.</summary>
public interface IMetaDataSource
{
    /// <summary>Human-readable id — used in parse-error messages (e.g. a filename).</summary>
    string Id { get; }
    /// <summary>Content-format hint — selects the parser.</summary>
    MetaDataFormat Format { get; }
    /// <summary>Resolve the raw content. May perform I/O.</summary>
    string Read();
}

/// <summary>Shared helpers for <see cref="IMetaDataSource"/> implementations.</summary>
internal static class MetaDataFormats
{
    /// <summary>Infer <see cref="MetaDataFormat"/> from a path/URI extension. Unknown → JSON.</summary>
    internal static MetaDataFormat InferFromExtension(string path)
    {
        string ext = System.IO.Path.GetExtension(path);
        return ext.Equals(".yaml", StringComparison.OrdinalIgnoreCase)
               || ext.Equals(".yml", StringComparison.OrdinalIgnoreCase)
            ? MetaDataFormat.Yaml
            : MetaDataFormat.Json;
    }
}

/// <summary>A metadata source backed by an in-memory string.</summary>
public sealed class InMemoryStringSource : IMetaDataSource
{
    private readonly string _content;
    public string Id { get; }
    public MetaDataFormat Format { get; }

    public InMemoryStringSource(string content, string id = "<inline>",
        MetaDataFormat format = MetaDataFormat.Json)
    {
        _content = content ?? throw new ArgumentNullException(nameof(content));
        Id = id ?? "<inline>";
        Format = format;
    }

    public string Read() => _content;
}
